### PR TITLE
ENH Anndata integration

### PR DESCRIPTION
**Pull request by BorisMuzellec**
This PR aims to support the [AnnData data structure](https://anndata.readthedocs.io/en/latest/) towards the integration of pydeseq2 to scverse (#15).

More precisely:

- The `DeseqDataSet` class now extends the `AnnData` class. As such, it inherits from all methods and attributes of `AnnData`, and parameters (such as size_factors and LFCs) are now stored in the relevant `AnnData` fields (e.g. `obsm` and `varm`),
- `DeseqStats` was updated accordingly,
- Pytests were updated,
- The sphinx documentation was adapted to handle the `AnnData` class inheritance, 
- The "Getting Started" notebooks from the sphinx-gallery were updated.

### CODE REVIEW

File: pydeseq2/dds.py [hunk 1]
```diff
@@ -88,63 +98,49 @@ class DeseqDataSet:
...
 
-    cooks : pandas.DataFrame
-        Cooks distances, used for outlier detection.
+    n_processes : int
+        Number of cpus to use for multiprocessing.
 
-    replaceable : pandas.Series
-        Whether counts are replaceable, i.e. if a given condition has enough samples.
+    non_zero : ndarray
+        Boolean mask of genes that have non-uniformly zero counts.
```
Comment (mandreux-owkin) [inline pydeseq2/dds.py:129]: Why not store it under `varm`? If I understand correctly, this variable would have the correct size and it is its purpose somehow?

File: pydeseq2/dds.py [hunk 2]
```diff
@@ -88,63 +98,49 @@ class DeseqDataSet:
...
-
-    size_factors : pandas.Series
-        DESeq normalization factors.
-
-    non_zero_genes : pandas.Index
-        Index of genes that have non-uniformly zero counts.
-
-    genewise_dispersions : pandas.Series
-        Initial estimates of gene counts dispersions.
+    X
```
Comment (mandreux-owkin) [inline pydeseq2/dds.py:101]: What is the type of `X`?

File: pydeseq2/dds.py [hunk 3]
```diff
@@ -88,63 +98,49 @@ class DeseqDataSet:
...
+        A ‘number of samples’ x ‘number of genes’ count data matrix.
 
-    trend_coeffs : pandas.Series
-        Coefficients of the trend curve: :math:`f(\mu) = \alpha_1/ \mu + a_0`.
+    obs
```
Comment (mandreux-owkin) [inline pydeseq2/dds.py:104]: I understand that `obs` is a dictionary, what is the type of each of its values? an `nd.array`? `pandas.Series`?

File: pydeseq2/dds.py [hunk 4]
```diff
@@ -88,63 +98,49 @@ class DeseqDataSet:
...
+        Key-indexed one-dimensional observations annotation of length 'number of
+        samples". Used to store design factors.
 
-    fitted_dispersions : pandas.Series
-        Genewise dispersions regressed on the trend curve.
+    var
+        Key-indexed one-dimensional gene-level annotation of length ‘number of genes’.
```
Comment (mandreux-owkin) [inline pydeseq2/dds.py:109]: Same question, what is the type of the entries of `var`?

File: pydeseq2/dds.py [hunk 5]
```diff
@@ -88,63 +98,49 @@ class DeseqDataSet:
...
-        MAP dispersions, after shrinkage towards the trend curve and before filtering.
+    obsm
+        Key-indexed multi-dimensional observations annotation of length
+        ‘number of samples’. Stores "design_matrix" and "size_factors", among others.
 
-    dispersions : pandas.Series
-        Final dispersion estimates, after filtering MAP outliers.
+    varm
+        Key-indexed multi-dimensional gene annotation of length ‘number of genes’.
+        Stores "dispersions" and "LFC", among others.
```
Comment (mandreux-owkin) [inline pydeseq2/dds.py:120]: What is the type of `varm`?

File: pydeseq2/dds.py [hunk 6]
```diff
@@ -88,63 +98,49 @@ class DeseqDataSet:
...
 
-    LFCs : pandas.DataFrame
-        Log-fold change and intercept parameters, in natural log scale.
+    layers
```
Comment (mandreux-owkin) [inline pydeseq2/dds.py:122]: What is the type of `layers`?

File: pydeseq2/dds.py [hunk 7]
```diff
@@ -288,12 +272,12 @@ def fit_genewise_dispersions(self):
                         batch_size=self.batch_size,
                     )(
                         delayed(fit_lin_mu)(
-                            counts=counts_nonzero[:, i],
-                            size_factors=size_factors,
+                            counts=self.X[:, i],
+                            size_factors=self.obsm["size_factors"],
                             design_matrix=X,
                             min_mu=self.min_mu,
```
Comment (mandreux-owkin) [inline pydeseq2/dds.py:278]: Cf above comment: these lines are confusing as `self.X` and `X` actually contain very different kinds of quantities

File: pydeseq2/dds.py [hunk 8]
```diff
@@ -332,48 +313,58 @@ def fit_genewise_dispersions(self):
...
-        self._genewise_converged = pd.Series(
-            l_bfgs_b_converged_, index=self.non_zero_genes
+
+        self.varm["genewise_dispersions"] = np.full(self.n_vars, np.NaN)
+        self.varm["genewise_dispersions"][self.non_zero] = np.clip(
+            dispersions_, self.min_disp, self.max_disp
         )
 
+        self.varm["_genewise_converged"] = np.full(self.n_vars, np.NaN)
+        self.varm["_genewise_converged"][self.non_zero] = np.array(l_bfgs_b_converged_)
```
Comment (mandreux-owkin) [inline pydeseq2/dds.py:337]: Not related to this PR, but this `np.array(l_bfgs_b_converged_)` does a copy of the underlying list (as they are not necessarily in contiguous areas), no? This may be inefficient, as it can be as large as the number of genes. Do you think there's a way to assign it beforehand and fill this array within `Parallel` somehow?

File: pydeseq2/dds.py [hunk 9]
```diff
@@ -461,16 +453,11 @@ def fit_MAP_dispersions(self):
...
 
-        # Exclude genes with all zeroes
-        counts_nonzero = self.counts.loc[:, self.non_zero_genes].values
-        num_genes = counts_nonzero.shape[1]
-
-        X = self.design_matrix.values
-        mu_hat = self._mu_hat.values
-        fit_disps = self.fitted_dispersions.loc[self.non_zero_genes].values
+        # Convert design matrix to numpy for speed
+        X = self.obsm["design_matrix"].values
```
Comment (mandreux-owkin) [inline pydeseq2/dds.py:460]: same comment as above, consider calling it `design_matrix_np`?

File: pydeseq2/dds.py [hunk 10]
```diff
@@ -522,16 +510,11 @@ def fit_LFC(self):
...
 
-        # Exclude genes with all zeroes
-        counts_nonzero = self.counts.loc[:, self.non_zero_genes].values
-        num_genes = counts_nonzero.shape[1]
-
-        X = self.design_matrix.values
-        size_factors = self.size_factors.values
-        disps = self.dispersions.loc[self.non_zero_genes].values
+        # Convert design matrix to numpy for speed
+        X = self.obsm["design_matrix"].values
```
Comment (mandreux-owkin) [inline pydeseq2/dds.py:517]: same comment, consider renaming `X` as `design_matrix_np`?

File: pydeseq2/dds.py [hunk 11]
```diff
@@ -542,49 +525,42 @@ def fit_LFC(self):
...
                     beta_tol=self.beta_tol,
                 )
-                for i in range(num_genes)
+                for i in self.non_zero_idx
             )
         end = time.time()
         print(f"... done in {end-start:.2f} seconds.\n")
 
-        MAP_lfcs_, mu_, hat_diagonals_, converged_ = zip(*res)
+        MLE_lfcs_, mu_, hat_diagonals_, converged_ = zip(*res)
```
Comment (mandreux-owkin) [inline pydeseq2/dds.py:540]: why this change in terminology?

File: pydeseq2/dds.py [hunk 12]
```diff
@@ -593,22 +569,39 @@ def calculate_cooks(self):
         """
 
         # Check that MAP dispersions are available. If not, compute them.
-        if not hasattr(self, "dispersions"):
+        if "dispersions" not in self.varm:
             self.fit_MAP_dispersions()
 
-        num_vars = self.design_matrix.shape[1]
+        num_vars = self.obsm["design_matrix"].shape[-1]
+        nz_data = self[:, self.non_zero_genes]
```
Comment (mandreux-owkin) [inline pydeseq2/dds.py:576]: maybe call it `nonzero_data` rather than `nz_data`? more explicit name (for 2s I hesitated with New Zealand...)

File: pydeseq2/dds.py [hunk 13]
```diff
@@ -593,22 +569,39 @@ def calculate_cooks(self):
...
+        self.layers["cooks"][:, self.non_zero] = (
             squared_pearson_res
             / num_vars
-            * (self._hat_diagonals / (1 - self._hat_diagonals) ** 2)
-        ).T
+            * (
+                nz_data.layers["_hat_diagonals"]
+                / (1 - nz_data.layers["_hat_diagonals"]) ** 2
+            )
+        )
```
Comment (mandreux-owkin) [inline pydeseq2/dds.py:604]: with the linting, the expression becomes a bit hard to follow. What about cutting it in pieces?

**Review by mandreux-owkin — COMMENTED** [review]
Congrats for the big refactoring work @BorisMuzellec ! The code seems correct to me. I have many cosmetic remarks that we can discuss in the comments. I think we'll be able to merge at the second pass.

File: docs/source/api/docstrings/pydeseq2.dds.DeseqDataSet.rst [hunk 14]
```diff
@@ -0,0 +1,21 @@
+﻿pydeseq2.dds.DeseqDataSet
```
[the remark this answers is not in the document]
Comment (BorisMuzellec) [reply docs/source/api/docstrings/pydeseq2.dds.DeseqDataSet.rst:1]: Hi @mandreux-owkin, I deactivated the autogeneration of rst files in [a3f9092](https://github.com/owkin/PyDESeq2/pull/66/commits/a3f9092f8809cdc2ef59af8a35aa9e939594bed0) to avoid the issue of documentation being created for attributes and methods inherited from anndata, which caused problems of readability and incompatible docstring formats.

Unless we find another solution and turn autogeneration on again, we should keep rst files in this PR.

File: docs/source/api/docstrings/pydeseq2.ds.DeseqStats.rst [hunk 15]
```diff
@@ -0,0 +1,20 @@
+﻿pydeseq2.ds.DeseqStats
```
[the remark this answers is not in the document]
Comment (BorisMuzellec) [reply docs/source/api/docstrings/pydeseq2.ds.DeseqStats.rst:1]: see above

File: docs/source/api/docstrings/pydeseq2.grid_search.rst [hunk 16]
```diff
@@ -0,0 +1,32 @@
+﻿pydeseq2.grid\_search
```
[the remark this answers is not in the document]
Comment (BorisMuzellec) [reply docs/source/api/docstrings/pydeseq2.grid_search.rst:1]: see above

File: docs/source/api/docstrings/pydeseq2.utils.rst [hunk 17]
```diff
@@ -0,0 +1,47 @@
+﻿pydeseq2.utils
```
[the remark this answers is not in the document]
Comment (BorisMuzellec) [reply docs/source/api/docstrings/pydeseq2.utils.rst:1]: see above

File: docs/source/api/docstrings/pydeseq2.preprocessing.rst [hunk 18]
```diff
@@ -0,0 +1,29 @@
+﻿pydeseq2.preprocessing
```
[the remark this answers is not in the document]
Comment (BorisMuzellec) [reply docs/source/api/docstrings/pydeseq2.preprocessing.rst:1]: see above

File: docs/source/api/index.rst [hunk 19]
```diff
@@ -7,10 +7,10 @@ PyDESeq2
 
 .. autosummary::
     :toctree: docstrings
-    :recursive:
```
[the remark this answers is not in the document]
Comment (BorisMuzellec) [reply docs/source/api/index.rst:10]: Indeed, it's not required here but it allows generating documents for modules and sub-modules recursively. I'll add it back.

File: docs/source/api/index.rst [hunk 20]
```diff
@@ -7,10 +7,10 @@ PyDESeq2
...
 
-    dds.DeseqDataSet
-    ds.DeseqStats
+    ~dds.DeseqDataSet
+    ~ds.DeseqStats
```
[the remark this answers is not in the document]
Comment (BorisMuzellec) [reply docs/source/api/index.rst:13]: In general it's used not to write the whole name. Not sure it's required here.

File: setup.py [hunk 21]
```diff
@@ -27,6 +27,7 @@
     author_email="boris.muzellec@owkin.com",
     packages=find_packages(exclude=["tests*"]),
     install_requires=[
+        "anndata==0.8.0",
```
[the remark this answers is not in the document]
Comment (BorisMuzellec) [reply setup.py:30]: I think it's better to require a fixed `anndata` version and check manually that everything keeps working when `anndata` is updated, to avoid `DeseqDatSet` being potentially broken by major `anndata` changes.

File: pydeseq2/dds.py (hunk 1 above)
Comment (BorisMuzellec) [reply pydeseq2/dds.py]: Done

File: pydeseq2/dds.py (hunk 2 above)
Comment (BorisMuzellec) [reply pydeseq2/dds.py:101]: The type of attributes inherited from `anndata` is a little complex (c.f. [their doc](https://anndata.readthedocs.io/en/latest/generated/anndata.AnnData.html#anndata.AnnData)).

I suggest leaving this blank for now and filling the gap using [autodoc-typehints](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_typehints) once we've merged the type hinting PR (as is done in AnnData's documentation).

File: pydeseq2/dds.py (hunk 3 above)
Comment (BorisMuzellec) [reply pydeseq2/dds.py:104]: cf above the [anndata doc](https://anndata.readthedocs.io/en/latest/generated/anndata.AnnData.html#anndata.AnnData)

File: pydeseq2/dds.py [hunk 22]
```diff
@@ -88,63 +98,49 @@ class DeseqDataSet:
...
-        Dispersion prior of genewise dispersions,
-        used for dispersion shrinkage towards the trend curve.
+    uns
+        Key-indexed unstructured annotation.
 
-    MAP_dispersions : pandas.Series
-        MAP dispersions, after shrinkage towards the trend curve and before filtering.
+    obsm
+        Key-indexed multi-dimensional observations annotation of length
+        ‘number of samples’. Stores "design_matrix" and "size_factors", among others.
```
[the remark this answers is not in the document]
Comment (BorisMuzellec) [reply pydeseq2/dds.py:116]: cf above

File: pydeseq2/dds.py (hunk 5 above)
Comment (BorisMuzellec) [reply pydeseq2/dds.py:120]: cf above

File: pydeseq2/dds.py (hunk 6 above)
Comment (BorisMuzellec) [reply pydeseq2/dds.py:122]: cf above

File: pydeseq2/dds.py (hunk 11 above)
Comment (BorisMuzellec) [reply pydeseq2/dds.py:539]: `MAP_lfcs` should be post-shrinkage LFCs. The previous name was a mistake

File: pydeseq2/dds.py [hunk 23]
```diff
@@ -630,70 +623,87 @@ def _fit_MoM_dispersions(self):
         """
 
         # Check that size_factors are available. If not, compute them.
-        if not hasattr(self, "size_factors"):
+        if "size_factors" not in self.obsm:
             self.fit_size_factors()
 
-        rde = fit_rough_dispersions(self.counts, self.size_factors, self.design_matrix)
-        mde = fit_moments_dispersions(self.counts, self.size_factors)
+        non_zero = ~(self.X == 0).all(axis=0)
```
[the remark this answers is not in the document]
Comment (BorisMuzellec) [reply pydeseq2/dds.py]: Indeed, good catch. Fixed this in [0d737b8](https://github.com/owkin/PyDESeq2/pull/66/commits/0d737b8cb5b8ff07bc6f925c0359a9cbf5cbe272)

File: pydeseq2/dds.py (hunk 10 above)
Comment (BorisMuzellec) [reply pydeseq2/dds.py]: done in [2bcc4dc](https://github.com/owkin/PyDESeq2/pull/66/commits/2bcc4dc7abc6d41951f4dcc70263b78a2b44ffdc)

File: pydeseq2/dds.py (hunk 9 above)
Comment (BorisMuzellec) [reply pydeseq2/dds.py]: done in [2bcc4dc](https://github.com/owkin/PyDESeq2/pull/66/commits/2bcc4dc7abc6d41951f4dcc70263b78a2b44ffdc)

File: pydeseq2/dds.py [hunk 24]
```diff
@@ -254,32 +241,29 @@ def fit_genewise_dispersions(self):
...
-        num_genes = counts_nonzero.shape[1]
+        self.non_zero = ~(self.X == 0).all(axis=0)
+        self.non_zero_idx = np.arange(self.n_vars)[self.non_zero]
+        self.non_zero_genes = self.var_names[self.non_zero]
 
         # Convert design_matrix to numpy for speed
-        X = self.design_matrix.values
-        rough_disps = self._rough_dispersions.loc[self.non_zero_genes].values
-        size_factors = self.size_factors.values
+        X = self.obsm["design_matrix"].values
```
[the remark this answers is not in the document]
Comment (BorisMuzellec) [reply pydeseq2/dds.py]: done in [2bcc4dc](https://github.com/owkin/PyDESeq2/pull/66/commits/2bcc4dc7abc6d41951f4dcc70263b78a2b44ffdc)

File: pydeseq2/dds.py [hunk 25]
```diff
@@ -254,32 +241,29 @@ def fit_genewise_dispersions(self):
...
         self._fit_MoM_dispersions()
 
         # Exclude genes with all zeroes
-        non_zero = ~(self.counts == 0).all()
-        self.non_zero_genes = non_zero[non_zero].index
-        counts_nonzero = self.counts.loc[:, self.non_zero_genes].values
-        num_genes = counts_nonzero.shape[1]
+        self.non_zero = ~(self.X == 0).all(axis=0)
+        self.non_zero_idx = np.arange(self.n_vars)[self.non_zero]
+        self.non_zero_genes = self.var_names[self.non_zero]
```
[the remark this answers is not in the document]
Comment (BorisMuzellec) [reply pydeseq2/dds.py]: Not sure, they're also used by `DeseqStats` (though also internally)

File: pydeseq2/dds.py (hunk 8 above)
Comment (BorisMuzellec) [reply pydeseq2/dds.py]: Done in [8f8322b](https://github.com/owkin/PyDESeq2/pull/66/commits/8f8322be117a08e9a40bb9d71671a241c88886a8)

File: pydeseq2/dds.py (hunk 13 above)
Comment (BorisMuzellec) [reply pydeseq2/dds.py]: done in [458d9d8](https://github.com/owkin/PyDESeq2/pull/66/commits/458d9d87fc25100e6bf44eefb1d828621cd61bd1)

File: pydeseq2/dds.py (hunk 11 above)
Comment (mandreux-owkin) [reply pydeseq2/dds.py:537]: ok thanks!

File: pydeseq2/dds.py (hunk 2 above)
Comment (mandreux-owkin) [reply pydeseq2/dds.py:101]: OK!

File: pydeseq2/dds.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (mandreux-owkin) [reply pydeseq2/dds.py]: ok, then let's keep them as is

File: docs/source/api/docstrings/pydeseq2.dds.DeseqDataSet.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (mandreux-owkin) [reply docs/source/api/docstrings/pydeseq2.dds.DeseqDataSet.rst:1]: thanks for the explanation! So be it.

File: setup.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (mandreux-owkin) [reply setup.py:30]: OK, let's keep this question on the side and move forward to merge quickly

**Review by mandreux-owkin — APPROVED** [review]
🎉 Let's merge! Thanks for all the changes @BorisMuzellec
